proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2521,6 +2439,445 @@ lemma possibleSwitchTo_all_invs_but_ct_not_inQ':
+                   apply fastforce
+                  apply (wpsimp wp: reply_unlink_tcb_valid_sched_action)
+                 apply wpsimp
+                apply (fastforce simp: vs_all_heap_simps pred_tcb_at_def obj_at_def)
+               apply clarsimp
+              apply (prop_tac "\<not> isReceive st'")
+               apply (case_tac st; clarsimp simp: isReceive_def)
+              apply (case_tac st
+                     ; clarsimp simp: isReceive_def
+                     ; (corressimp corres: restart_thread_if_no_fault_corres[simplified isNothing_def]
+                        , fastforce simp: obj_at_def))
+             apply (wpsimp wp: gts_wp)
+            apply (wpsimp wp: gts_wp')
+           apply (clarsimp simp: vs_all_heap_simps obj_at_def is_tcb_def)
+          apply clarsimp
+         apply (intro hoare_vcg_conj_lift_left_fix; (solves \<open>wpsimp wp: gts_wp\<close>)?)
+           apply (wpsimp wp: restart_thread_if_no_fault_tcb_sts_of_other
+                             reply_unlink_tcb_tcb_sts_of_other gts_wp)
+         apply (rule hoare_weaken_pre
+               , rule cancel_all_ipc_loop_body_reply_unlink_ts_pred_other
+               , fastforce)
+        apply (intro hoare_vcg_conj_lift_left_fix; (solves wpsimp)?)
+        apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_not_runnable'_other, simp)
+       apply (clarsimp simp: pred_conj_def)
+       apply (intro hoare_vcg_conj_lift_left_fix; (solves \<open>wpsimp wp: gts_wp\<close>)?)
+       apply (wpsimp wp: cancel_all_ipc_loop_body_valid_sched)
+      apply (clarsimp simp: pred_conj_def)
+      apply (intro hoare_vcg_conj_lift_left_fix; (solves wpsimp)?)
+       apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_weak_sch_act_wf, simp)
+      apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_valid_queues, simp)
+     apply fastforce+
+  done
+
+crunches replyUnlink
+  for sch_act_not[wp]: "\<lambda>s. sch_act_not t s"
+  and typ_at'[wp]: "\<lambda>s. P (typ_at' T p s)"
+  (wp: crunch_wps)
+
+lemmas replyUnlink_typ_ats[wp] = typ_at_lifts[OF replyUnlink_typ_at']
+
+(* FIXME RT: is there an equivalent to typ_at_lifts? This fact doesn't seem to be produced
+             by reply_unlink_tcb_typ_at *)
+crunches reply_unlink_tcb
+  for ep_at[wp]: "\<lambda>s. ep_at ep_ptr s"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma in_send_ep_queue_TCBBlockedSend:
+  "\<lbrakk>kheap s epptr = Some (Endpoint (Structures_A.endpoint.SendEP queue)); t \<in> set queue; invs s\<rbrakk>
+   \<Longrightarrow> (epptr, TCBBlockedSend) \<in> state_refs_of s t"
+  apply (prop_tac "valid_ep (Structures_A.endpoint.SendEP queue) s")
+   apply (fastforce simp: valid_objs_def valid_obj_def dest!: invs_valid_objs)
+  apply (clarsimp simp: state_refs_of_def valid_ep_def split: option.splits)
+  apply (intro conjI impI allI; (fastforce simp: obj_at_def)?)
+  apply (prop_tac "(t, EPSend) \<in> state_refs_of s epptr")
+   apply (clarsimp simp: state_refs_of_def)
```

Comment:
Nitpick/suggestion. In this case, I like to use a comma to put the one-line proof directly after the prop_tac. I think it reads more clearly.